### PR TITLE
admin: Add ban endpoint.

### PR DESCRIPTION
**Pull request by JoeGruffins**

### CODE REVIEW

File: server/admin/types.go [hunk 1]
```diff
@@ -42,10 +43,20 @@ func (at *APITime) MarshalJSON() ([]byte, error) {
...
+		return nil
+	}
+	// Parenthesis are included in b and must be removed.
+	t, err := time.Parse(RFC3339Milli, string(b[1:len(b)-1]))
 	if err != nil {
 		return nil
 	}
 	at.Time = t
 	return nil
 }
```
Comment (JoeGruffins) [inline server/admin/types.go:56]: I believe this was creating a bug where the time always unmarshalled to nil, or zero time.

File: server/auth/auth.go [hunk 2]
```diff
@@ -626,11 +626,26 @@ func (auth *AuthManager) RequestWithTimeout(user account.AccountID, msg *msgjson
...
+		log.Error(err)
+	}
+}
+
+// Ban signals that a user has broken a rule of community conduct, and that
+// their account should be penalized. It returns an error and is used by remote
+// callers.
+func (auth *AuthManager) Ban(user account.AccountID, rule account.Rule) error {
+	return auth.penalize(user, rule)
+}
```
Comment (JoeGruffins) [inline server/auth/auth.go:642]: I just needed the error to return to the remote caller, so I broke Penalize up. Penalize is still exactly the same but with an extra debug log for successes.

File: server/auth/auth.go [hunk 3]
```diff
@@ -626,11 +626,26 @@ func (auth *AuthManager) RequestWithTimeout(user account.AccountID, msg *msgjson
...
+
+// penalize signals that a user has broken a rule of community conduct, and that
+// their account should be penalized.
+func (auth *AuthManager) penalize(user account.AccountID, rule account.Rule) error {
 	if auth.anarchy {
-		log.Infof("user %v penalized for rule %v, but not enforcing it", user, rule)
-		return
+		return fmt.Errorf("user %v penalized for rule %v, but not enforcing it", user, rule)
```
Comment (JoeGruffins) [inline server/auth/auth.go:648]: Since we weren't able to penalized as expected, this is an error now.

File: server/account/account.go [hunk 4]
```diff
@@ -102,4 +102,7 @@ const (
 	// PreimageReveal means an account failed to respond with a valid preimage
 	// for their order during epoch processing.
 	PreimageReveal
+	// Admin means an account was banned by the operator through the admin
+	// interface.
+	Admin
```
Comment (JoeGruffins) [inline server/account/account.go:107]: I couldn't think of a better name.

**Comment by JoeGruffins** [comment]
I see the four penalization rules are outlined in the spec. Should banning from the admin server be one of those four? I guess that's more correct than arbitrary banning for an unspecified reason.

File: server/account/account.go (hunk 4 above)
Comment (JoeGruffins) [reply server/account/account.go]: I think adding an arbitrary rule is a bad idea. A client should not be banned for unspecified reasons.

File: server/auth/auth.go [hunk 5]
```diff
@@ -626,11 +626,26 @@ func (auth *AuthManager) RequestWithTimeout(user account.AccountID, msg *msgjson
...
+// Ban signals that a user has broken a rule of community conduct, and that
+// their account should be penalized. It returns an error and is used by remote
+// callers.
+func (auth *AuthManager) Ban(user account.AccountID, rule account.Rule) error {
+	return auth.penalize(user, rule)
+}
+
+// penalize signals that a user has broken a rule of community conduct, and that
+// their account should be penalized.
+func (auth *AuthManager) penalize(user account.AccountID, rule account.Rule) error {
```
Comment (JoeGruffins) [inline server/auth/auth.go:646]: I think there should be a check in this method that rule is one of account.Rule, and something to be penalized for, ie rule > 0 && rule < account.MaxRule.

File: server/auth/auth.go (hunk 2 above)
Comment (chappjc) [reply server/auth/auth.go:642]: I don't want two nearly identical functions.  Let's just have Penalize return and error and update the interfaces.  Callers can ignore the error if they want.

File: server/admin/server_test.go [hunk 6]
```diff
@@ -597,7 +601,7 @@ func TestSuspend(t *testing.T) {
 	}
 
 	wantFinal = encode.UnixTimeMilli(tMsFuture + 1)
-	if suspRes.SuspendTime.Equal(wantFinal) {
+	if !suspRes.SuspendTime.Equal(wantFinal) {
```
Comment (chappjc) [inline server/admin/server_test.go:604]: oops

**Comment by JoeGruffins** [comment]
I was looking at suspend yesterday, and arguments are sent as tokens `suspend?t=EPOCH-MS&persist=BOOL`, should the rule variable mirror that?

**Comment by chappjc** [comment]
Since we're not strictly following REST naming conventions (we using verbs instead of POST for simplicity), I can't rightly be too strict on this issue.  Although I suppose `/ban/{acctID}` is one more step away convention.  Let's go with your suggestion.

File: server/account/account.go [hunk 7]
```diff
@@ -89,7 +89,7 @@ type Rule uint8
 const (
 	// NoRule indicates that no rules have been broken. This may be an invalid
 	// value in some contexts.
-	NoRule = iota
+	NoRule Rule = iota
```
Comment (JoeGruffins) [inline server/account/account.go:92]: I think this was intended, but not sure.

File: server/admin/api.go [hunk 8]
```diff
@@ -180,3 +184,43 @@ func (s *Server) apiAccountInfo(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, "rule not specified", http.StatusBadRequest)
+		return
+	}
+	ruleInt, err := strconv.Atoi(ruleStr)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("bad rule: %v", err), http.StatusBadRequest)
+		return
+	}
+	if ruleInt < 1 || ruleInt >= int(account.MaxRule) {
+		http.Error(w, "bad rule: not known or not punishable", http.StatusBadRequest)
```
Comment (chappjc) [inline server/admin/api.go:211]: Was about to push but the no golangci-lint complains about these two unneeded Sprintf calls (gosimple).  Fixed.

**Pull request merged**